### PR TITLE
refactor: use nullable `ParserInput`

### PR DESCRIPTION
**Pull request by magnus-madsen**

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/ast/SourceLocation.scala [hunk 1]
```diff
@@ -40,18 +42,19 @@ object SourceLocation {
...
+  * @param input        the nullable parser input.
   * @param source       the source input.
   * @param isReal       true if real location, false if synthetic location.
   * @param beginLine    the line number where the entity begins.
   * @param beginCol     the column number where the entity begins.
   * @param endLine      the line number where the entity ends.
   * @param endCol       the column number where the entity ends.
   */
-case class SourceLocation(input: Option[ParserInput], source: Source, isReal: Boolean, beginLine: Int, beginCol: Short, endLine: Int, endCol: Short) {
+case class SourceLocation(@Nullable input: ParserInput, source: Source, isReal: Boolean, beginLine: Int, beginCol: Short, endLine: Int, endCol: Short) {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/ast/SourceLocation.scala:57]: Can we make `input` private so the null ability cannot leak?

File: main/src/ca/uwaterloo/flix/language/ast/SourcePosition.scala [hunk 2]
```diff
@@ -3,21 +3,23 @@ package ca.uwaterloo.flix.language.ast
...
 /**
   * A class that represent a physical source position inside a source input.
   *
   * @param line  the line number.
   * @param col   the column number.
-  * @param input the parser input.
+  * @param input the nullable parser input.
   */
-case class SourcePosition(source: Source, line: Int, col: Short, input: Option[ParserInput]) {
+case class SourcePosition(source: Source, line: Int, col: Short, @Nullable input: ParserInput) {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/ast/SourcePosition.scala:22]: Same

File: main/src/ca/uwaterloo/flix/language/ast/SourceLocation.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/SourceLocation.scala:57]: Case classes cannot have private fields.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/SourceLocation.scala:57]: Ah they can. But I can only make it private for source location.

**Comment by magnus-madsen** [comment]
@mlutze Fixed

**Review by mlutze — APPROVED** [review]

**Pull request merged**